lib/typetraits.h

```diff
@@ -0,0 +1,23 @@
+// SPDX-FileCopyrightText: 2022-2024, Alejandro Colomar <alx@kernel.org>
+// SPDX-License-Identifier: BSD-3-Clause
+
+
+#ifndef SHADOW_INCLUDE_LIB_TYPETRAITS_H_
+#define SHADOW_INCLUDE_LIB_TYPETRAITS_H_
+
+
+#include <config.h>
+
+#include "sizeof.h"
+
+
+#define is_unsigned(x)  (((typeof(x)) -1) > 1)
+#define is_signed(x)    (((typeof(x)) -1) < 1)
+
+#define stype_max(T)    ((T) (((((T) 1 << (WIDTHOF(T) - 2)) - 1) << 1) + 1))
```

Comment:
Call me chicken, but could you add a simple testcase for just one example type, enforcing that stype_max and utype_max do the right thing for, say, an 8 bit type?

Reply:
I'm preparing a test.